### PR TITLE
Recruitment web app - administrator - CRUD employee #18

### PR DESCRIPTION
**Pull request by jjjsss3**
Issue: https://github.com/HopeeSolutionsRuby/demo-app/issues/18
Description: 
- Migrate create model employee
- Admin Employee Controller for CRUD
- View for CRUD

Note (optional): 
- Have used pagy for pagination, ransack for search, turbo-rails for ajax
-  Have not yet shown the avatar on the view (can select to create and update)

### CODE REVIEW

**Review by tunh-dev1 — APPROVED** [review]
LGTM!

File: app/controllers/admins/employees_controller.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module Admins
+  class EmployeesController < ApplicationController
+    layout 'admin/layout'
+    before_action :set_employee, only: %i[edit update destroy show]
+
+    def index
+      @url = admins_employees_path
```
Comment (KhoaNDAHopee) [inline app/controllers/admins/employees_controller.rb]: No need to assign this variable to views

File: app/controllers/admins/employees_controller.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      @queries = 'id_or_first_name_or_last_name_or_address_or_account_email_or_account_phone_cont'
```
Comment (KhoaNDAHopee) [inline app/controllers/admins/employees_controller.rb]: No need to assign this variable to views

File: app/controllers/admins/employees_controller.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      @q = Employee.includes(:account).ransack(params[:q])
+      @employees = @q.result(distinct: true)
```
Comment (KhoaNDAHopee) [inline app/controllers/admins/employees_controller.rb]: No need to assign this variable to views

File: app/controllers/admins/employees_controller.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      @pagy, @records = pagy(@employees)
+    end
+
+    def show
+      employees = employee_current_before_after_get
```
Comment (KhoaNDAHopee) [inline app/controllers/admins/employees_controller.rb]: Don't load 3 employees every go to detail page, only load 1

File: app/models/concerns/capitalizable.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Capitalizable
+  extend ActiveSupport::Concern
+
+  class_methods do
+    def capitalize_attribute(attribute)
```
Comment (KhoaNDAHopee) [inline app/models/concerns/capitalizable.rb]: No need

File: app/models/concerns/employee_activities.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module EmployeeActivities
+  extend ActiveSupport::Concern
+  include Capitalizable
+
+  included do
+    after_create :send_email_register
+    capitalize_attribute :first_name
```
Comment (KhoaNDAHopee) [inline app/models/concerns/employee_activities.rb]: No need

File: app/models/concerns/employee_activities.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    capitalize_attribute :last_name
```
Comment (KhoaNDAHopee) [inline app/models/concerns/employee_activities.rb]: No need

File: app/models/concerns/employee_activities.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    def self.create_with_account(params)
```
Comment (KhoaNDAHopee) [inline app/models/concerns/employee_activities.rb]: Move it to services https://viblo.asia/p/rails-service-objects-guide-bWrZnNkmZxw

File: app/models/concerns/employee_activities.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      [employee, nil]
+    rescue ActiveRecord::RecordInvalid => e
+      Rails.logger.debug e.message
+      [nil, e.message]
+    end
+  end
+
+  private
+
+  def send_email_register
```
Comment (KhoaNDAHopee) [inline app/models/concerns/employee_activities.rb]: No need

File: app/views/admins/employees/_table.html.erb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+<table class="table table-hover">
+  <thead>
+    <tr>
+      <th scope="col"><%= sort_link(@q, :id, "#", class: 'text-dark text-decoration-none')  %></th>
```
Comment (KhoaNDAHopee) [inline app/views/admins/employees/_table.html.erb]: ``one') %>`` check spacing, check another.

File: app/views/admins/employees/_table_line.html.erb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<tr>
+  <th scope="row"><%= employee.id%></th>
```
Comment (KhoaNDAHopee) [inline app/views/admins/employees/_table_line.html.erb]: ``id%>`` spacing

File: app/views/admins/employees/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    <%= link_to new_admins_employee_path, class: 'btn btn-primary', data: { turbo_frame: 'container-frame', turbo_action: 'replace' } do %>
+      <i class="fa-solid fa-plus"></i>
+    <% end %>
+    <%= link_to admins_employees_path, class: "btn btn-secondary", data: { turbo_frame: 'container-frame', turbo_action: 'replace' } do %>
+      <i class="fa-solid fa-list-ul"></i>
+    <% end %>
+  </div>
+</div>
+<%= form_with model: @employee, url: admins_employee_path do |form| %>
+  <%# First name, last name %>
```
Comment (KhoaNDAHopee) [inline app/views/admins/employees/edit.html.erb]: No need to use ruby comments

File: app/views/admins/employees/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    </div>
+    <div class="flex-fill ms-1">
+      <%= form.label :password_confirm, class: "form-label"%><br>
+      <%= form.password_field :password_confirmation, class: "form-control"%>
+    </div>
+  </div>
+  <%# Avatar %>
+  <div class="mb-3">
+    <%= form.label :avatar, for:"formFileLg", class:"form-label" %>
+    <%= form.file_field :avatar, class:"form-control mb-2", id:"formFileSm"%>
```
Comment (KhoaNDAHopee) [inline app/views/admins/employees/new.html.erb]: - ``id:"formFileSm"`` spacing
- Follow naming conventions for id, class =>`` form-file-sm``, check another

File: app/views/shared/_pagination.html.erb [hunk 14]
```diff
@@ -0,0 +1 @@
+<%== pagy_bootstrap_nav(@pagy) if @pagy.pages > 0 %>
```
Comment (KhoaNDAHopee) [inline app/views/shared/_pagination.html.erb]: - ``<%==``
- Don't show pagination if page = 1

File: app/views/shared/_search_box.html.erb [hunk 15]
```diff
@@ -1,16 +1,19 @@
 <div class="search-form">
-  <%= form_with do |form| %>
-    <i class="fa-solid fa-magnifying-glass"></i>
-    <%= form.text_field :text, class: "form-control form-input", placeholder: "Search any thing..." %> 
+  <% if @q %>
+    <%= search_form_for @q, url: @url, remote: true, data: { turbo_frame: 'container-frame' } do |f| %>
```
Comment (KhoaNDAHopee) [inline app/views/shared/_search_box.html.erb]: ``@url`` Use view helpers instead get instance variable from controller

File: app/views/shared/admin/_left_menu.html.erb [hunk 16]
```diff
@@ -34,7 +34,7 @@
 
  <%# Accounts %>
   <li class="nav-item py-2">
-    <a class="nav-link row d-flex" href="#" role="button" data-toggle="collapse" data-target="#collapse_accounts" aria-expanded="true" aria-controls="collapse2">
+    <a class="nav-link row d-flex" role="button" data-toggle="collapse" data-target="#collapse_accounts" aria-expanded="true" aria-controls="collapse2">
```
Comment (KhoaNDAHopee) [inline app/views/shared/admin/_left_menu.html.erb]: Use view helpers

File: app/views/shared/admin/_left_menu.html.erb [hunk 17]
```diff
@@ -58,7 +58,7 @@
 
   <%# Employers %>
   <li class="nav-item py-2">
-    <a class="nav-link row d-flex" href="#" role="button" data-toggle="collapse" data-target="#collapse_employers" aria-expanded="true" aria-controls="collapse2">
+    <a class="nav-link row d-flex" role="button" data-toggle="collapse" data-target="#collapse_employers" aria-expanded="true" aria-controls="collapse2">
```
Comment (KhoaNDAHopee) [inline app/views/shared/admin/_left_menu.html.erb]: Use view helpers

File: app/views/shared/admin/_left_menu.html.erb [hunk 18]
```diff
@@ -82,7 +82,7 @@
 
   <%# Employees %>
   <li class="nav-item py-2">
-    <a class="nav-link row d-flex" href="#" role="button" data-toggle="collapse" data-target="#collapse_employees" aria-expanded="true" aria-controls="collapse2">
+    <a class="nav-link row d-flex" role="button" data-toggle="collapse" data-target="#collapse_employees" aria-expanded="true" aria-controls="collapse2">
```
Comment (KhoaNDAHopee) [inline app/views/shared/admin/_left_menu.html.erb]: Use view helpers

File: config/initializers/pagy.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+require 'pagy/extras/bootstrap'
+Pagy::DEFAULT[:items] = 10
+Pagy::DEFAULT[:size]  = [1,4,4,1]
```
Comment (KhoaNDAHopee) [inline config/initializers/pagy.rb]: Check code style

**Review by KhoaNDAHopee — CHANGES_REQUESTED** [review]

**Review by DucNguyen0812 — APPROVED** [review]
LGTM!

**Pull request closed without merging**